src/compute/src/render/context.rs

```diff
@@ -44,100 +44,199 @@ use crate::typedefs::{
+    }
+
+    /// Returns the frontier before which updates should not be emitted.
     ///
     /// We *must* apply it to sinks, to ensure correct outputs.
     /// We *should* apply it to sources and imported traces, because it improves performance.
-    pub as_of_frontier: Antichain<T>,
-    /// Frontier after which updates should not be emitted.
+    fn as_of(&self) -> &Antichain<mz_repr::Timestamp> {
+        &self.static_().as_of
+    }
+
+    /// Returns the frontier after which updates should not be emitted.
+    ///
     /// Used to limit the amount of work done when appropriate.
-    pub until: Antichain<T>,
-    /// Bindings of identifiers to collections.
-    pub bindings: BTreeMap<Id, CollectionBundle<S, T>>,
-    /// A token that operators can probe to know whether the dataflow is shutting down.
-    pub(super) shutdown_token: ShutdownToken,
-    /// Specification for rendering linear joins.
-    pub(super) linear_join_spec: LinearJoinSpec,
-    pub(super) enable_specialized_arrangements: bool,
+    fn until(&self) -> &Antichain<mz_repr::Timestamp> {
+        &self.static_().until
+    }
+
+    /// Returns the specification for rendering linear joins.
+    fn linear_join_spec(&self) -> LinearJoinSpec {
+        self.static_().linear_join_spec
+    }
+
+    /// Returns whether arrangement specialization is enabled.
+    fn enable_specialized_arrangements(&self) -> bool {
+        self.static_().enable_specialized_arrangements
+    }
+
+    /// Returns a shutdown-aware error logger to be used by rendered operators.
+    fn error_logger(&self) -> ErrorLogger {
+        ErrorLogger::new(self.shutdown_token().clone(), self.debug_name().into())
+    }
+
+    /// Creates a new dataflow region with the same timestamp.
+    fn region<F, R>(&mut self, name: &str, func: F) -> R
+    where
+        F: FnOnce(&mut ChildContext<Self>) -> R,
+    {
+        self.scope().clone().region_named(name, |region| {
+            let mut inner = ChildContext {
+                scope: region.clone(),
+                bindings: Default::default(),
+                parent: self,
+            };
+            func(&mut inner)
+        })
+    }
+}
+
+/// Static context information about a dataflow.
+pub(super) struct StaticContext {
+    debug_name: String,
+    dataflow_id: usize,
+    as_of: Antichain<mz_repr::Timestamp>,
+    until: Antichain<mz_repr::Timestamp>,
+    linear_join_spec: LinearJoinSpec,
+    enable_specialized_arrangements: bool,
 }
 
-impl<S: Scope> Context<S>
+/// The root context within which a dataflow is rendered.
+pub(super) struct RootContext<S>
 where
-    S::Timestamp: Lattice + Refines<mz_repr::Timestamp> + Columnation,
+    S: Scope,
+    S::Timestamp: RenderTimestamp,
 {
-    /// Creates a new empty Context.
-    pub fn for_dataflow_in<Plan>(
-        dataflow: &DataflowDescription<Plan, CollectionMetadata>,
+    scope: S,
+    bindings: BTreeMap<Id, CollectionBundle<S>>,
+    static_: StaticContext,
```

Comment:
Could we rename this to `static_context`? I find the trailing `_` odd as it might indicate "don't look at this."